### PR TITLE
[SPARK-9772] [PySpark] [ML] Add Python API for ml.feature.VectorSlicer

### PR DESCRIPTION
**Pull request by yanboliang**
Add Python API for ml.feature.VectorSlicer.

### CODE REVIEW

**Comment by yanboliang** [comment]
Jenkins, test this please.

**Comment by jkbradley** [comment]
test this please

**Comment by jkbradley** [comment]
reviewing now

File: python/pyspark/ml/feature.py [hunk 1]
```diff
@@ -950,6 +950,92 @@ class VectorIndexerModel(JavaModel):
...
+
+    The output vector will order features with the selected indices first (in the order given),
+    followed by the selected names (in the order given).
+
+    >>> from pyspark.mllib.linalg import Vectors
+    >>> df = sqlContext.createDataFrame([
+    ...     (Vectors.dense([-2.0, 2.3, 0.0, 0.0, 1.0]),),
+    ...     (Vectors.dense([0.0, 0.0, 0.0, 0.0, 0.0]),),
+    ...     (Vectors.dense([0.6, -1.1, -3.0, 4.5, 3.3]),)], ["features"])
+    >>> vs = VectorSlicer(inputCol="features", outputCol="expected", indices=[1, 4])
```
Comment (jkbradley) [inline python/pyspark/ml/feature.py]: Rename "expected" to "sliced" since this is an example

File: python/pyspark/ml/feature.py [hunk 2]
```diff
@@ -950,6 +950,92 @@ class VectorIndexerModel(JavaModel):
...
+    >>> vs.transform(df).head().expected
+    DenseVector([2.3, 1.0])
+    """
+
+    # a placeholder to make it appear in the generated doc
+    indices = Param(Params._dummy(), "indices", "An array of indices to select features from " +
+                    "a vector column. There can be no overlap with `names`.")
+    names = Param(Params._dummy(), "names", "An array of feature names to select features from " +
+                  "a vector column. These names must be specified by ML " +
+                  "`org.apache.spark.ml.attribute.Attribute`s. There can be no overlap with " +
```
Comment (jkbradley) [inline python/pyspark/ml/feature.py]: No need for backtick within strings; the backtick is only for formatting generated API docs.  (here and elsewhere)

File: python/pyspark/ml/feature.py [hunk 3]
```diff
@@ -950,6 +950,92 @@ class VectorIndexerModel(JavaModel):
...
+                  "`indices`.")
+
+    @keyword_only
+    def __init__(self, inputCol=None, outputCol=None, indices=[], names=[]):
```
Comment (jkbradley) [inline python/pyspark/ml/feature.py]: In Python, we should avoid using mutable values ```[]``` as defaults.  Let's use None.  The Scala API should take care of the defaults.

File: python/pyspark/ml/feature.py [hunk 4]
```diff
@@ -950,6 +950,92 @@ class VectorIndexerModel(JavaModel):
...
+        """
+        super(VectorSlicer, self).__init__()
+        self._java_obj = self._new_java_obj("org.apache.spark.ml.feature.VectorSlicer", self.uid)
+        self.indices = Param(self, "indices", "An array of indices to select features from " +
+                             "a vector column. There can be no overlap with `names`.")
+        self.names = Param(self, "names", "An array of feature names to select features from " +
+                           "a vector column. These names must be specified by ML " +
+                           "`org.apache.spark.ml.attribute.Attribute`s. There can be no overlap " +
+                           "with `indices`.")
+        self._setDefault(indices=[], names=[])
```
Comment (jkbradley) [inline python/pyspark/ml/feature.py]: so no defaults here

**Comment by jkbradley** [comment]
One more comment: need to add VectorSlicer to list ```__all__``` at top of file

**Comment by yanboliang** [comment]
Jenkins, test this please.

**Comment by jkbradley** [comment]
LGTM.  Merging with master.  Thank you!

**Pull request closed without merging**